Objective-C/CBLDocument.mm

```diff
@@ -198,80 +201,102 @@ - (NSUInteger)countByEnumeratingWithState: (NSFastEnumerationState *)state
+    CBL_LOCK(self.lock) {
+        return _database;
+    }
+}
+
+
 - (void) setDatabase: (CBLDatabase *)database {
-    _database = database;
-    _c4db = _database.c4db;
+    CBL_LOCK(self.lock) {
+        _database = database;
+        _c4db = _database.c4db;
+    }
 }
 
 
 - (BOOL) isEmpty {
-    return _dict.isEmpty;
+    return self.dict.isEmpty;
 }
 
 
 - (BOOL) save: (NSError**)outError {
-    return [self saveWithConflictResolver: self.effectiveConflictResolver
-                                 deletion: NO
-                                    error: outError];
+    CBL_LOCK(self.lock) {
+        return [self saveWithConflictResolver: self.effectiveConflictResolver
+                                     deletion: NO
+                                        error: outError];
+    }
 }
 
 
 - (BOOL) deleteDocument: (NSError**)outError {
-    return [self saveWithConflictResolver: self.effectiveConflictResolver
-                                 deletion: YES
-                                    error: outError];
+    CBL_LOCK(self.lock) {
+        return [self saveWithConflictResolver: self.effectiveConflictResolver
+                                     deletion: YES
+                                        error: outError];
+    }
 }
 
 
 - (BOOL) purge: (NSError**)outError {
-    assert(_database && _c4db);
-    
-    if (!self.exists) {
-        return createError(kCBLStatusNotFound, outError);
-    }
-    
-    C4Transaction transaction(_c4db);
-    if (!transaction.begin())
-        return convertError(transaction.error(),  outError);
-    
-    C4Error err;
-    if (c4doc_purgeRevision(self.c4Doc.rawDoc, C4Slice(), &err) >= 0) {
-        if (c4doc_save(self.c4Doc.rawDoc, 0, &err)) {
-            // Save succeeded; now commit:
-            if (!transaction.commit())
-                return convertError(transaction.error(), outError);
-            
-            // Reset:
-            [self setC4Doc: nil];
-            return YES;
+    CBL_LOCK(self.lock) {
+        assert(_database && _c4db);
+        
+        if (!self.exists) {
+            return createError(kCBLStatusNotFound, outError);
+        }
+        
+        C4Transaction transaction(_c4db);
+        if (!transaction.begin())
+            return convertError(transaction.error(),  outError);
+        
+        C4Error err;
+        if (c4doc_purgeRevision(self.c4Doc.rawDoc, C4Slice(), &err) >= 0) {
+            if (c4doc_save(self.c4Doc.rawDoc, 0, &err)) {
+                // Save succeeded; now commit:
+                if (!transaction.commit())
+                    return convertError(transaction.error(), outError);
+                
+                // Reset:
+                [self setC4Doc: nil];
+                return YES;
+            }
         }
+        return convertError(err, outError);
     }
-    return convertError(err, outError);
 }
 
 
 #pragma mark - PRIVATE
 
 
+- (CBLDictionary*) dict {
+    CBL_LOCK(self.lock) {
+        return _dict;
```

Comment:
This isn't safe, for the same reason as the -database method. Let the compiler synthesize this.